File: pg-index-health-spring-boot-demo/src/main/java/io/github/mfvanek/pg/index/health/demo/service/DbMigrationGeneratorService.java
/*
 * Copyright (c) 2019-2024. Ivan Vakhrushev and others.
 * https://github.com/mfvanek/pg-index-health-demo
 *
 * Licensed under the Apache License 2.0
 */

package io.github.mfvanek.pg.index.health.demo.service;

import io.github.mfvanek.pg.generator.DbMigrationGenerator;
import io.github.mfvanek.pg.health.checks.common.DatabaseCheckOnCluster;
import io.github.mfvanek.pg.index.health.demo.dto.ForeignKeyMigrationResponse;
import io.github.mfvanek.pg.model.constraint.ForeignKey;
import io.github.mfvanek.pg.model.context.PgContext;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.sql.Connection;
import java.sql.SQLException;
import java.sql.Statement;
import java.util.List;
import javax.sql.DataSource;

@Slf4j
@RequiredArgsConstructor
@Service
@Transactional(readOnly = true)
public class DbMigrationGeneratorService {

    private final DataSource dataSource;
    private final DbMigrationGenerator<ForeignKey> dbMigrationGenerator;
    private final DatabaseCheckOnCluster<ForeignKey> foreignKeysNotCoveredWithIndex;
    private final PgContext pgContext;

    public ForeignKeyMigrationResponse generateMigrationsWithForeignKeysChecked() {
        final List<ForeignKey> keysBefore = getForeignKeysFromDb();
        final List<String> migrations = generateMigrations(keysBefore);
        runGeneratedMigrations(migrations);
        final List<ForeignKey> keysAfter = getForeignKeysFromDb();
        if (!keysAfter.isEmpty()) {
            throw new IllegalStateException("There should be no foreign keys not covered by the index");
        }
        return new ForeignKeyMigrationResponse(keysBefore, keysAfter, migrations);
    }

    List<ForeignKey> getForeignKeysFromDb() {
        return foreignKeysNotCoveredWithIndex.check(pgContext);
    }

    private List<String> generateMigrations(final List<ForeignKey> foreignKeys) {
        final List<String> generatedMigrations = dbMigrationGenerator.generate(foreignKeys);
        log.info("Generated migrations: {}", generatedMigrations);
        return generatedMigrations;
    }

    private void runGeneratedMigrations(final List<String> generatedMigrations) {
        try (Connection connection = dataSource.getConnection()) {
            for (final String migration : generatedMigrations) {
                try (Statement statement = connection.createStatement()) {
                    statement.execute(migration);
                }
            }
        } catch (SQLException e) {
            log.error("Error running migration", e);
        }
    }
}

File: pg-index-health-spring-boot-kotlin-demo/src/main/kotlin/io/github/mfvanek/pg/index/health/demo/kotlin/service/DbMigrationGeneratorService.kt
/*
 * Copyright (c) 2019-2025. Ivan Vakhrushev and others.
 * https://github.com/mfvanek/pg-index-health-demo
 *
 * Licensed under the Apache License 2.0
 */

package io.github.mfvanek.pg.index.health.demo.kotlin.service

import io.github.mfvanek.pg.generator.DbMigrationGenerator
import io.github.mfvanek.pg.health.checks.common.DatabaseCheckOnCluster
import io.github.mfvanek.pg.index.health.demo.kotlin.dto.ForeignKeyMigrationResponse
import io.github.mfvanek.pg.index.health.demo.kotlin.exception.MigrationException
import io.github.mfvanek.pg.index.health.demo.kotlin.mapper.ForeignKeyMapper
import io.github.mfvanek.pg.model.constraint.ForeignKey
import io.github.mfvanek.pg.model.context.PgContext
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.sql.SQLException
import javax.sql.DataSource

/**
 * Service for generating database migrations.
 *
 * @property dataSource Data source for database connections
 * @property dbMigrationGenerator Generator for creating migration scripts
 * @property foreignKeysNotCoveredWithIndex Check for foreign keys without indexes
 * @property pgContext PostgreSQL context
 */
@Service
@Transactional(readOnly = true)
class DbMigrationGeneratorService(
    private val dataSource: DataSource,
    private val dbMigrationGenerator: DbMigrationGenerator<ForeignKey>,
    private val foreignKeysNotCoveredWithIndex: DatabaseCheckOnCluster<ForeignKey>,
    private val pgContext: PgContext,
    private val foreignKeyMapper: ForeignKeyMapper
) {

    private val logger = LoggerFactory.getLogger(DbMigrationGeneratorService::class.java)

    /**
     * Generates migrations for foreign keys and validates the result.
     *
     * @return response containing foreign keys before and after migration, and generated migrations
     * @throws MigrationException if there are still foreign keys without indexes after migration
     */
    fun generateMigrationsWithForeignKeysChecked(): ForeignKeyMigrationResponse {
        val keysBefore = getForeignKeysFromDb()
        val migrations = generateMigrations(keysBefore)
        runGeneratedMigrations(migrations)
        val keysAfter = getForeignKeysFromDb()
        if (keysAfter.isNotEmpty()) {
            throw MigrationException("There should be no foreign keys not covered by some index")
        }
        return ForeignKeyMigrationResponse(
            keysBefore.map { foreignKeyMapper.toForeignKeyDto(it) },
            keysAfter.map { foreignKeyMapper.toForeignKeyDto(it) },
            migrations
        )
    }

    /**
     * Gets foreign keys from the database that are not covered with indexes.
     *
     * @return list of foreign keys without indexes
     */
    internal fun getForeignKeysFromDb(): List<ForeignKey> {
        return foreignKeysNotCoveredWithIndex.check(pgContext)
    }

    /**
     * Generates migration scripts for the given foreign keys.
     *
     * @param foreignKeys list of foreign keys to generate migrations for
     * @return list of generated migration scripts
     */
    private fun generateMigrations(foreignKeys: List<ForeignKey>): List<String> {
        val generatedMigrations = dbMigrationGenerator.generate(foreignKeys)
        logger.info("Generated migrations: {}", generatedMigrations)
        return generatedMigrations
    }

    /**
     * Runs the generated migration scripts against the database.
     *
     * @param generatedMigrations list of migration scripts to execute
     */
    private fun runGeneratedMigrations(generatedMigrations: List<String>) {
        try {
            dataSource.connection.use { connection ->
                for (migration in generatedMigrations) {
                    executeMigration(connection, migration)
                }
            }
        } catch (e: SQLException) {
            logger.error("Error getting connection", e)
        }
    }

    /**
     * Executes a single migration script.
     *
     * @param connection database connection
     * @param migration migration script to execute
     */
    private fun executeMigration(connection: java.sql.Connection, migration: String) {
        try {
            connection.createStatement().use { statement ->
                statement.execute(migration)
            }
        } catch (e: SQLException) {
            logger.error("Error running migration", e)
        }
    }
}
